projects/extension/ai/openai.py

```diff
@@ -17,58 +18,95 @@ def get_openai_api_key(plpy, api_key_name: Optional[str] = None) -> str:
-        "select pg_catalog.current_setting('ai.openai_base_url', true) as base_url"
+        "SELECT pg_catalog.current_setting('ai.openai_base_url', true) AS base_url"
     )
     if len(r) == 0:
         return None
     return r[0]["base_url"]
 
 
-def make_client(
-    plpy,
-    api_key: Optional[str] = None,
-    api_key_name: Optional[str] = None,
-    base_url: Optional[str] = None,
-) -> openai.Client:
+def make_async_client(
+        plpy,
+        api_key: Optional[str] = None,
+        api_key_name: Optional[str] = None,
+        organization: Optional[str] = None,
+        base_url: Optional[str] = None,
+        timeout: Optional[float] = None,
+        max_retries: Optional[int] = None,
+        default_headers: Optional[Dict[str, str]] = None,
+        default_query: Optional[Dict[str, Any]] = None,
+        http_client: Optional[Any] = None,
+        _strict_response_validation: Optional[bool] = None
+) -> openai.AsyncOpenAI:
     if api_key is None:
         api_key = get_openai_api_key(plpy, api_key_name)
     if base_url is None:
         base_url = get_openai_base_url(plpy)
-    return openai.Client(api_key=api_key, base_url=base_url)
-
-
-def list_models(
-    plpy,
-    api_key: Optional[str] = None,
-    api_key_name: Optional[str] = None,
-    base_url: Optional[str] = None,
-) -> Generator[tuple[str, datetime, str], None, None]:
-    client = make_client(plpy, api_key, api_key_name, base_url)
-    from datetime import datetime, timezone
-
-    for model in client.models.list():
-        created = datetime.fromtimestamp(model.created, timezone.utc)
-        yield model.id, created, model.owned_by
-
-
-def embed(
-    plpy,
-    model: str,
-    input: Union[str, list[str], list[int]],
-    api_key: Optional[str] = None,
-    api_key_name: Optional[str] = None,
-    base_url: Optional[str] = None,
-    dimensions: Optional[int] = None,
-    user: Optional[str] = None,
-) -> Generator[tuple[int, list[float]], None, None]:
-    client = make_client(plpy, api_key, api_key_name, base_url)
-    args = {}
-    if dimensions is not None:
-        args["dimensions"] = dimensions
-    if user is not None:
-        args["user"] = user
-    response = client.embeddings.create(input=input, model=model, **args)
-    if not hasattr(response, "data"):
-        return None
-    for obj in response.data:
-        yield obj.index, obj.embedding
+
+    client_kwargs = prepare_kwargs({
+        "api_key": api_key,
+        "organization": organization,
+        "base_url": base_url,
+        "timeout": timeout,
+        "max_retries": max_retries,
+        "default_headers": default_headers,
+        "default_query": default_query,
+        "http_client": http_client,
+        "_strict_response_validation": _strict_response_validation
+    })
+
+    return openai.AsyncOpenAI(**client_kwargs)
+
+def get_or_create_client(plpy, GD: Dict[str, Any], api_key: str = None, api_key_name: str = None, base_url: str = None) -> Any:
```

Comment:
Do you have any number showing that creating the client is expensive (and thus worth it to store in GD)?. Does this allow connection reuse or something? And if it's the latter then how/when do connections get closed? Is there a keepalive timeout.

Storing the client in GD seems like a good amount of complexity and I'd like to find out what we are gaining/loosing for it.

Reply:
Yup, benchmarks here: https://github.com/timescale/pgai/issues/116#issuecomment-2421594920

Also note, that there is a known issue with the 2nd (and 3rd...etc) call to the client for the api has some extra 20ms delay that doesn't happen when I have this code running outside of a pl/python environment (noted in the thread above). I really should have mentioned that directly in the PR. will edit to mention that fact that it needs to be identified. Once that is fixed, the benchmark numbers should look much better.

Even with the above issue, this is still much faster, and lower CPU than the original implementation where we recreate the client.

Note specifically the CPU reduction. Recreating the client is heavy on CPU, I know this from past projects but the benchmarks also bare this out.

I believe the connection is closed after the request is complete, and the client becomes ready for the next call. If the request is cancelled early, we attempt to kill things gracefully.